Revert "Fix MacOS build conflicting with system Python 3"

rdesgroppes opened the pull request:
### What does this PR do? / Motivation
The present change is a follow-up of #39563, where it was found that:
> We need to have a default value here because these unit tests are also called from `macos` runners  but [`system` is patched as `linux`](https://github.com/DataDog/datadog-agent/pull/39563/files#diff-407acb07e932b50660fc4999612d12a05d6046a1716a53125e1010050c898cebR33). So even though the unit tests are run in macos runners(run on EC2) they try to access vault and dont have K8 `POD_NAMESPACE` env var set.

It turns out the temporary fix introduced by #5010 in 2020 is no longer needed:
- upstream bug report: https://github.com/pypa/virtualenv/issues/1643
- upstream fix: https://github.com/pypa/virtualenv/pull/1648
- released with `virtualenv` v20.0.5: https://virtualenv.pypa.io/en/latest/changelog.html#v20-0-5-2020-02-21

This therefore reverts commit d16a8a4da3baadc1ada4636513476e5b3df169cb: "[omnibus] Fix MacOS build conflicting with system Python 3 (#5010)", and adjusts consequences accordingly:
 - the `POD_NAMESPACE` environment variable must be present (no default) for `linux`, and only for `linux`,
- patches in tests get tailored to the very needs, with 3 `platforms` to cover implying 3 "sub" tests,
- covering `win32` unearthes unmatched command patterns due to suffixes (`.bat`, `.exe`) being used for certain commands (like `aws.exe` and `omnibus.bat`, unlike `git`) which requires further adaptation, especially to understand which patterns are not found.

rdesgroppes committed:
Revert "Fix MacOS build conflicting with system Python 3"

The present change is a follow-up of #39563, where it was found that:
> We need to have a default value here because these unit tests are also
> called from `macos` runners  but [`system` is patched as
> `linux`](https://github.com/DataDog/datadog-agent/pull/39563/files#diff-407acb07e932b50660fc4999612d12a05d6046a1716a53125e1010050c898cebR33).
> So even though the unit tests are run in macos runners(run on EC2)
> they try to access vault and dont have K8 `POD_NAMESPACE` env var set.

It turns out the temporary fix introduced by #5010 in 2020 is no longer
needed:
- upstream bug report:
  https://github.com/pypa/virtualenv/issues/1643
- upstream fix:
  https://github.com/pypa/virtualenv/pull/1648
- released with `virtualenv` v20.0.5:
  https://virtualenv.pypa.io/en/latest/changelog.html#v20-0-5-2020-02-21

This therefore reverts commit d16a8a4da3baadc1ada4636513476e5b3df169cb:
"[omnibus] Fix MacOS build conflicting with system Python 3", and
adjusts consequences accordingly:
- the `POD_NAMESPACE` environment variable must be present (no default)
  for `linux`, and only for `linux`,
- patches in tests get tailored to the very needs. (3 `platforms` => 3
  "sub" tests)

rdesgroppes commented on tasks/unit_tests/omnibus_tests.py:
```diff
@@ -30,7 +32,6 @@ def _run_calls_to_string(mock_calls):
     return '\n'.join(commands_run)
 
 
-@mock.patch('sys.platform', 'linux')
```
The problems I see is that:
1. `@mock.patch('sys.platform', 'linux')` prevents from covering `darwin` and `win32` code paths,
2. contributors are confused leading them to workarounds `os.environ.get('POD_NAMESPACE', 'gitlab-runner')` instead of the intended `os.environ["POD_NAMESPACE"]` - so was I.

Unless I'm mistaken, tests pass on Windows. Do you foresee only a subset of tests was run and we'll face a problem after merging to the mainline?

rdesgroppes replied:
Done, now running these tests for all platforms irrelevantly from the host OS.

aiuto commented on tasks/omnibus.py:
```diff
@@ -36,14 +36,7 @@ def omnibus_run_task(ctx, task, target_project, base_dir, env, log_level="info",
-            omnibus = "unset __PYVENV_LAUNCHER__ && bundle exec omnibus"
-
+        omnibus = f"bundle exec omnibus{".bat" if sys.platform == "win32" else ""}"
         cmd = "{omnibus} {task} {project_name} --log-level={log_level} {overrides}"
```
I see the pattern here is to create command lines that get passed to ctx.run().
Is that really what we do?  It's so much safer to pass a vector of args, and let ctx run properly quote per OS when it does the fork/exec.  If we have that option, we could start migrating to it.

alopezz replied:
This may work but the [type annotation of the run function](https://docs.pyinvoke.org/en/stable/api/runners.html#invoke.runners.Runner.run) asks for a string.

rdesgroppes commented:
> probably a rebase error here

Indeed, thanks @paulcacheux!

rdesgroppes commented on tasks/unit_tests/omnibus_tests.py:
```diff
@@ -165,17 +187,20 @@ def test_cache_hit_with_corruption(self):
 
-        omnibus.build(self.mock_ctx)
+        with mock.patch.dict("os.environ") as env:
+            del env["OMNIBUS_GIT_CACHE_DIR"]
```
... whereas this only mutates the patched `os.environ` (test execution stack).

alopezz commented on tasks/unit_tests/omnibus_tests.py:
```diff
@@ -40,27 +41,45 @@ def _run_calls_to_string(mock_calls):
+        def decorator(test_func):
+            @functools.wraps(test_func)
+            def wrapper(self, *args, **kwargs):
+                for platform, get_dd_api_key_env in {
+                    "darwin": {"AGENT_API_KEY_ORG2": "agent-api-key"},
+                    "linux": {"AGENT_API_KEY_ORG2": "agent-api-key", "POD_NAMESPACE": "pod-ns"},
+                    "win32": {"API_KEY_ORG2": "api-key"},
+                }.items():
```
This is a point where I miss pytest 😬; interesting way to solve this regardless.

rdesgroppes replied:
I also miss `pytest`, but [nose](https://nose.readthedocs.io/en/latest/) even more.

alopezz commented on tasks/unit_tests/omnibus_tests.py:
```diff
@@ -40,27 +41,45 @@ def _run_calls_to_string(mock_calls):
+                    with (
+                        self.subTest(platform=platform),
+                        mock.patch("sys.platform", platform),
+                        mock.patch.dict(os.environ, get_dd_api_key_env) if calls_get_dd_api_key else nullcontext(),
```
Question: would there be any drawbacks to making this patching unconditional?

rdesgroppes replied:
Quite the other way around. Let me shrink it.

rdesgroppes replied:
✅ Done.

alopezz reviewed (commented):
Looks good to me, my only real concern is the one about the assertion on `bundle exec omnibus` when win32 should have `bundle exec omnibus.py`.

rdesgroppes committed:
Make the decorator's patching unconditional

This applies review suggestion:
https://github.com/DataDog/datadog-agent/pull/39633/#discussion_r2276017803

rdesgroppes committed:
Prevent `MockContextRaising` reuse across subtests

This addresses following review comment:
https://github.com/DataDog/datadog-agent/pull/39633/#discussion_r2276055212

rdesgroppes commented on tasks/unit_tests/omnibus_tests.py:
```diff
@@ -76,6 +95,7 @@ def assertRunLines(self, line_patterns):
             f'Failed to match pattern {line_patterns}.',
         )
 
+    @_for_each_platform()
```
Yes, there's was a subtle bug... the following method turns out to be only invoked prior to executing a `test_*` method, but not between sub-tests:
```
    def setUp(self):
        self.mock_ctx = MockContextRaising(run={})
```
✅ Please find a fix in the 3rd commit.

rdesgroppes committed:
Restore in-order pattern matching (with precise reporting)

alopezz reviewed (approved):
Thanks!
